dialogue-core/src/main/java/com/palantir/dialogue/core/RefreshingChannelFactory.java

```diff
@@ -73,7 +82,26 @@ private AlwaysThrowingChannel(String serviceName) {
         }
     }
 
-    public interface ChannelFactory {
-        Channel create(ClientConfiguration conf);
+    @VisibleForTesting
+    static final class RefreshingChannel implements Channel {
+        private final Supplier<Channel> channelSupplier;
+
+        private RefreshingChannel(Supplier<Channel> channelSupplier) {
+            this.channelSupplier = channelSupplier;
+        }
+
+        /**
+         * Returns a Channel which will be built by applying the {@code channelFactory} to whatever the latest config
+         * provided by the {@code confSupplier} is. Avoids invoking the channelFactory if the config hasn't changed.
+         */
+        static <T> Channel create(Supplier<T> confSupplier, Function<T, Channel> channelFactory) {
+            Supplier<Channel> channelSupplier = new MemoizingComposingSupplier<>(confSupplier, channelFactory);
```

Comment:
FWIW I don't think this is the ideal long-term solution, because we've it a bug with a similar memoizing supplier that contained a `synchronized` block internally in sls-logging.  Just making the minimum changes necessary to unblock my simulations.

Reply:
Mind adding a comment along these lines?